# Download genomes of listed species, avoiding duplicates

import urllib2
import string
import os.path

def parse_line(line,species,entries):

	if line.startswith('>'): # new fasta entry
		name = line[line.find(" ")+1:].lower()[:-1]
		if name.find(species) != -1: # matches species
			if name in entries or (name.find('complete genome') != -1 and entries != []): # duplicate name or duplicate 'complete genome'
				print "\tSkipping duplicate {0}".format(name)
				return False
			else: # valid new fasta that matches species
				entries.append(name)
				print "Added {0}".format(name)
				return '>' + species.replace (" ", "_") + " " + line[1:] #add name to beginning of ID line
		else: # does not match species
			print "\tSkipping {0}".format(name)
			return False
	elif line.startswith('<'): # xml instead of fasta
		print "ERROR: FASTA EMPTY PAGE"
		return False

	return line # normal line of fasta

def get_genome(species_orig):
	species = species_orig.lower()
	filename = species.replace (" ", "_").replace('/',"") + '.mfa'

	if os.path.isfile(filename) and os.path.getsize(filename) > 200: # don't re-run if file exists and is sensible size
		return
	else:
		print "SPECIES: {0}".format(species_orig)

	mfa = open(filename, 'w') # new fasta file for every species
	mfa.seek(0)

	if 'ID_list' in globals():
		# species ID
		tax_id = ID_list[species_list.index(species_orig)]

		# get species name
		try:
			page_tax = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id={0}'.format(tax_id))
		except:
			print "FAILED TO GET TAXONOMY PAGE FOR {0}".format(species)
			return 0

		for line in page_tax:
			line = string.replace(line,"\t","") # unfortunately this xml page uses spaces
			if line.startswith('    <ScientificName>'):
				species = line[20:-18].lower()
				print 'Actual species name: {0}'.format(species)
				break

	# get genome ID
	page_genome = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=genome&term=' + urllib2.quote(species))
	genome_id = ""
	for line in page_genome:
		line = string.replace(line,"\t","")
		if line.startswith("<Id>"):
			if genome_id != "":
				print "Duplicate IDs for {0}; skipping.".format(species)
				return
			genome_id = line[4:-6]

	# get query key and web env
	page_nucleotide = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=genome&db=nuccore&id={0}&term=gene+in+chromosome[prop]+OR+gene+in+plasmid[prop]+OR+wgs[prop]&cmd=neighbor_history'.format(genome_id))
	for line in page_nucleotide:
		line = string.replace(line,"\t","")
		if line.startswith("<QueryKey>"):
			query_key = line[10:-12]
		if line.startswith("<WebEnv>"):
			web_env = line[8:-10]

	# get fastas
	try:
		page_fasta = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&query_key={0}&WebEnv={1}&rettype=fasta&retmode=text'.format(query_key,web_env))
		print  'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&query_key={0}&WebEnv={1}&rettype=fasta&retmode=text'.format(query_key,web_env)
	except:
		print "FAILED TO GET FASTA PAGE FOR {0}".format(species)
		return 0

	# copy fastas, trying to skip duplicates
	fasta = ""
	skip = False
	entries = []
	for line in page_fasta:
		if not skip or line.startswith('>'): # if skip, avoid entire loop until next new record
			skip = False
			nl = parse_line(line,species,entries)
			if nl:
				fasta = fasta + nl
			else:
				skip = True
	'''
	# get query key and web env for plasmid
	page_nucleotide = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=genome&db=nuccore&id=' + str(genome_id) + '&term=gene+in+plasmid[prop]&cmd=neighbor_history')
	for line in page_nucleotide:
		line = string.replace(line,"\t","")
		if line.startswith("<QueryKey>"):
			query_key = line[10:-12]
		if line.startswith("<WebEnv>"):
			web_env = line[8:-10]

	# get fasta for plasmid
	page_fasta = urllib2.urlopen('http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&query_key=' + query_key + '&WebEnv=' + web_env + '&rettype=fasta&retmode=text')

	# copy plasmid fastas, trying to skip duplicates
	fasta = fasta + "\n"
	skip = False
	entries = []
	for line in page_fasta:
		if not skip or line.startswith('>'): # if skip, avoid entire loop until next new record
			skip = False
			nl = parse_line(line,species,entries)
			if nl:
				fasta = fasta + nl
			else:
				skip = True
	'''
	fasta = fasta + "\n"
	mfa.write(fasta) # write fastas to file
	mfa.truncate()
	mfa.close()

i100_species_list = [
'Psychrobacter cryohalolentis K5',
'Mycobacterium sp. JLS',
'Synechococcus elongatus PCC 7942',
'Bacillus cereus ATCC 14579',
'Lactococcus lactis subsp. lactis Il1403',
'Wigglesworthia glossinidia endosymbiont of Glossina brevipalpis',
'Pseudomonas putida W619',
'Streptococcus pneumoniae R6',
'Bordetella bronchiseptica RB50',
'Lactobacillus delbrueckii subsp. bulgaricus ATCC BAA-365',
'Clostridium novyi NT',
'Staphylococcus aureus subsp. aureus JH1',
'Borrelia garinii PBi',
'Salmonella enterica subsp. enterica serovar Paratyphi B str. SPB7',
'Erwinia tasmaniensis',
'Staphylococcus aureus subsp. aureus MW2',
'Thiobacillus denitrificans ATCC 25259',
'Rhodopseudomonas palustris BisA53',
'Bacillus anthracis str. Ames',
'Streptococcus agalactiae 2603V/R',
'Burkholderia phymatum STM815',
'Campylobacter concisus 13826',
'Alkaliphilus metalliredigens QYMF',
'Corynebacterium efficiens YS-314',
'Streptococcus pyogenes MGAS5005',
'Ochrobactrum anthropi ATCC 49188',
'Prochlorococcus marinus str. MIT 9313',
'Rhodopseudomonas palustris CGA009',
'Corynebacterium urealyticum DSM 7109',
'Thermus thermophilus HB8',
'Acinetobacter baumannii SDF',
'Neisseria gonorrhoeae FA 1090',
'Hyphomonas neptunium ATCC 15444',
'Staphylococcus aureus subsp. aureus str. Newman',
'Haloquadratum walsbyi DSM 16790',
'Prochlorococcus marinus str. NATL1A',
'Clostridium beijerinckii NCIMB 8052',
'Chromobacterium violaceum ATCC 12472',
'Chlamydia trachomatis L2b/UCH-1/proctitis',
'Cytophaga hutchinsonii ATCC 33406',
'Alkalilimnicola ehrlichii MLHE-1', #fixed spelling
'Bacillus thuringiensis str. Al Hakam',
'Clostridium botulinum B1 str. Okra',
'Methanococcus maripaludis C7',
'Sodalis glossinidius str. \'morsitans\'',
'Cyanothece sp. ATCC 51142',
'Lactococcus lactis subsp. cremoris SK11',
'Haemophilus somnus 129PT',
'Clostridium thermocellum ATCC 27405',
'Neisseria meningitidis MC58',
'Ignicoccus hospitalis KIN4/I',
'Escherichia coli E24377A',
'Thermosipho melanesiensis BI429',
'Lawsonia intracellularis PHE/MN1-00',
'Coxiella burnetii RSA 493',
'Yersinia pestis CO92',
'Bacillus cereus E33L',
'Candidatus Blochmannia pennsylvanicus str. BPEN',
'Idiomarina loihiensis L2TR',
'Syntrophomonas wolfei subsp. wolfei str. Goettingen',
'Bacillus cereus ATCC 10987',
'Escherichia coli APEC O1',
'Mycobacterium tuberculosis F11',
'Bacillus clausii KSM-K16',
'Lactobacillus salivarius UCC118',
'Mycobacterium bovis AF2122/97',
'Nostoc punctiforme PCC 73102',
'Shewanella sp. MR-7',
'Burkholderia pseudomallei 668',
'Rickettsia prowazekii str. Madrid E',
'Haemophilus influenzae Rd KW20',
'Pseudomonas stutzeri A1501',
'Treponema denticola ATCC 35405',
'Ralstonia eutropha H16',
'Mycobacterium avium subsp. paratuberculosis K-10',
'Listeria welshimeri serovar 6b str. SLCC5334',
'Bacillus halodurans C-125',
'Geobacter sulfurreducens PCA',
'Chlamydophila pneumoniae CWL029',
'Francisella tularensis subsp. tularensis FSC198',
'Staphylococcus aureus subsp. aureus NCTC 8325',
'Mycobacterium marinum M',
'Bacillus subtilis subsp. subtilis str. 168',
'Staphylococcus aureus subsp. aureus MRSA252',
'Sinorhizobium meliloti 1021',
'Streptococcus pyogenes MGAS315',
'Ureaplasma parvum serovar 3 str. ATCC 700970',
'Chlamydophila caviae GPIC',
'Anabaena variabilis ATCC 29413',
'Psychromonas ingrahamii 37',
'Synechococcus elongatus PCC 6301',
'Bacteroides thetaiotaomicron VPI-5482',
'Staphylococcus aureus subsp. aureus Mu3',
'Leptospira borgpetersenii serovar Hardjo-bovis JB197',
'Herpetosiphon aurantiacus DSM 785', #new strain name
'Buchnera aphidicola BCc', #new strain name
'Rhodococcus jostii RHA1', #new species name
'Escherichia coli str. K-12 substr. W3110', #added -
'Escherichia coli str. K-12 substr. MG1655', #added -
'Chlorobium luteolum DSM 273', #new genus
'pseudomonas fluorescens pf0-1',
'Rhodopseudomonas palustris HaA2',
'Burkholderia ambifaria AMMD']

i100_inactive_ID_list = [ #NCBI taxonomy id
316058,
288000,
269798,
155920,
264732,
155919,
269484,
266117,
292415,
315749,
269483,
351627,
324602,
290402,
165597,
332415,
59920,
335284,
316056,
94122,
60481,
292414,
340099,
339671,
339670,
290315,
319795,
290400,
266940,
265072,
323098,
279238,
324925,
296591,
342610,
316057,
317655,
326298,
203124,
187272,
240292,
290397,
290399,
322710,
331271,
331272,
269482,
203119,
272564,
207559,
262543,
106370,
298653,
269799,
321967,
324831,
351348,
323259,
323097,
323261,
335283,
323848,
196162,
338963,
338966,
351746,
205918,
338969,
316055,
269796,
326297,
325240,
318167,
319224,
323850,
286604,
335543,
293826,
205913,
321955,
290317,
159087,
333146,
205914,
272622,
203120,
156889,
318586,
278197,
319225,
74546,
290512,
331678,
203122,
351745,
335541,
317025,
266265,
290318,
290398,
333849,
321956,
266779,
259564,
269797,
205922,
259536,
1140,
269800,
203123,
272943,
387344,
322159]

#i400
species_list = [
"Anaeromyxobacter sp. Fw109-5",
"Maricaulis maris MCS10",
"Flavobacterium johnsoniae UW101",
"Mycoplasma synoviae 53",
"Neorickettsia sennetsu str. Miyayama",
"Rickettsia typhi str. Wilmington",
"Pseudomonas putida KT2440",
"Heliobacterium modesticaldum Ice1",
"Ureaplasma parvum serovar 3 str. ATCC 27815",
"Silicibacter pomeroyi DSS-3",
"Mycobacterium sp. KMS",
"Nostoc sp. PCC 7120",
"Rhodopirellula baltica SH 1",
"Pyrobaculum calidifontis JCM 11548",
"Sulfolobus acidocaldarius DSM 639",
"Ralstonia eutropha JMP134",
"Rickettsia akari str. Hartford",
"Nocardioides sp. JS614",
"Shigella dysenteriae Sd197",
"Burkholderia pseudomallei 1710b",
"Methanopyrus kandleri AV19",
"Desulfococcus oleovorans Hxd3",
"Pseudomonas stutzeri A1501",
"Helicobacter acinonychis str. Sheeba",
"Bradyrhizobium sp. BTAi1",
"Lactobacillus fermentum IFO 3956",
"Oenococcus oeni PSU-1",
"Lactobacillus plantarum WCFS1",
"Leuconostoc mesenteroides subsp. mesenteroides ATCC 8293",
"uncultured methanogenic archaeon RC-I",
"Staphylococcus aureus subsp. aureus USA300",
"Delftia acidovorans SPH-1",
"Mycoplasma capricolum subsp. capricolum ATCC 27343",
"Marinobacter aquaeolei VT8",
"Arthrobacter aurescens TC1",
"Mycobacterium marinum M",
"Bacillus cereus E33L",
"Actinobacillus pleuropneumoniae L20",
"Brucella canis ATCC 23365",
"Mycoplasma mobile 163K",
"Ehrlichia chaffeensis str. Arkansas",
"Salinispora arenicola CNS-205",
"Shewanella amazonensis SB2B",
"Cupriavidus taiwanensis",
"Ehrlichia ruminantium str. Welgevonden",
"Cyanothece sp. ATCC 51142",
"Pseudoalteromonas atlantica T6c",
"Synechococcus sp. PCC 7002",
"Aquifex aeolicus VF5",
"Vibrio harveyi ATCC BAA-1116",
"Acaryochloris marina MBIC11017",
"Staphylococcus epidermidis RP62A",
"Polaromonas sp. JS666",
"Candidatus Pelagibacter ubique HTCC1062",
"Roseiflexus castenholzii DSM 13941",
"Thermobifida fusca YX",
"Fusobacterium nucleatum subsp. nucleatum ATCC 25586",
"Pseudomonas fluorescens Pf-5",
"Shewanella baltica OS195",
"Borrelia afzelii PKo",
"Aeromonas salmonicida subsp. salmonicida A449",
"Thermoanaerobacter pseudethanolicus ATCC 33223",
"Oceanobacillus iheyensis HTE831",
"Burkholderia mallei NCTC 10247",
"Halorhodospira halophila SL1",
"Mesorhizobium sp. BNC1",
"Paracoccus denitrificans PD1222",
"Dehalococcoides ethenogenes 195",
"Thermosipho melanesiensis BI429",
"Mycobacterium leprae TN",
"Methanococcus aeolicus Nankai-3",
"Ralstonia metallidurans CH34",
"Saccharophagus degradans 2-40",
"Bifidobacterium adolescentis ATCC 15703",
"Xanthomonas axonopodis pv. citri str. 306",
"Coxiella burnetii Dugway 5J108-111",
"Streptococcus pneumoniae CGSP14",
"Bacillus weihenstephanensis KBAB4",
"Granulibacter bethesdensis CGDNIH1",
"Lactobacillus sakei subsp. sakei 23K",
"Aster yellows witches'-broom phytoplasma AYWB",
"Pseudoalteromonas haloplanktis TAC125",
"Roseiflexus sp. RS-1",
"Buchnera aphidicola str. APS (Acyrthosiphon pisum)",
"Mycobacterium vanbaalenii PYR-1",
"Lactobacillus acidophilus NCFM",
"Acidothermus cellulolyticus 11B",
"Hyphomonas neptunium ATCC 15444",
"Lactococcus lactis subsp. cremoris SK11",
"Natronomonas pharaonis DSM 2160",
"Shewanella denitrificans OS217",
"Pyrobaculum arsenaticum DSM 13514",
"Parvibaculum lavamentivorans DS-1",
"Dechloromonas aromatica RCB",
"Nitrobacter winogradskyi Nb-255",
"Methanococcus vannielii SB",
"Mannheimia succiniciproducens MBEL55E",
"Burkholderia thailandensis E264",
"Mycobacterium gilvum PYR-GCK",
"Gramella forsetii KT0803",
"Geobacter uraniireducens Rf4",
"Clostridium kluyveri DSM 555",
"Ignicoccus hospitalis KIN4/I",
"Geobacter sulfurreducens PCA",
"Bartonella henselae str. Houston-1",
"Corynebacterium efficiens YS-314",
"Chloroflexus aurantiacus J-10-fl",
"Staphylococcus haemolyticus JCSC1435",
"Gluconacetobacter diazotrophicus PAl 5",
"Acidiphilium cryptum JF-5",
"Lactobacillus delbrueckii subsp. bulgaricus ATCC BAA-365",
"Bacillus pumilus SAFR-032",
"Rickettsia prowazekii str. Madrid E",
"Thermotoga lettingae TMO",
"Myxococcus xanthus DK 1622",
"Geobacillus thermodenitrificans NG80-2",
"Nostoc punctiforme PCC 73102",
"Bacteroides vulgatus ATCC 8482",
"Methanobrevibacter smithii ATCC 35061",
"Polynucleobacter necessarius STIR1",
"Acidovorax sp. JS42",
"Streptococcus thermophilus LMD-9",
"Streptococcus suis 05ZYH33",
"Aeromonas hydrophila subsp. hydrophila ATCC 7966",
"Shewanella woodyi ATCC 51908",
"Nitrobacter hamburgensis X14",
"Haemophilus somnus 2336",
"Burkholderia cenocepacia MC0-3",
"Silicibacter sp. TM1040",
"Azoarcus sp. EbN1",
"Shewanella loihica PV-4",
"Psychrobacter arcticus 273-4",
"Pyrococcus abyssi GE5",
"Sulfurovum sp. NBC37-1",
"Neisseria meningitidis MC58",
"Bartonella tribocorum CIP 105476",
"Listeria welshimeri serovar 6b str. SLCC5334",
"Shewanella pealeana ATCC 700345",
"Lactobacillus reuteri F275",
"Clostridium difficile 630",
"Haemophilus influenzae PittGG",
"Borrelia garinii PBi",
"Geobacillus kaustophilus HTA426",
"Halobacterium salinarum R1",
"Leuconostoc citreum KM20",
"Syntrophomonas wolfei subsp. wolfei str. Goettingen",
"Microcystis aeruginosa NIES-843",
"Thermosynechococcus elongatus BP-1",
"Thermus thermophilus HB27",
"Thermoanaerobacter sp. X514",
"Leptospira interrogans serovar Lai str. 56601",
"Porphyromonas gingivalis ATCC 33277",
"Shewanella frigidimarina NCIMB 400",
"Nitrosomonas europaea ATCC 19718",
"Leifsonia xyli subsp. xyli str. CTCB07",
"Mycoplasma agalactiae PG2",
"Picrophilus torridus DSM 9790",
"Chlorobium tepidum TLS",
"Trichodesmium erythraeum IMS101",
"Lawsonia intracellularis PHE/MN1-00",
"Anaeromyxobacter dehalogenans 2CP-C",
"Corynebacterium urealyticum DSM 7109",
"Pseudomonas aeruginosa PA7",
"Halobacterium sp. NRC-1",
"Nanoarchaeum equitans Kin4-M",
"Rickettsia bellii RML369-C",
"Geobacter lovleyi SZ",
"Photobacterium profundum SS9",
"Sulfurimonas denitrificans DSM 1251",
"Methanocorpusculum labreanum Z",
"Thermoanaerobacter tengcongensis MB4",
"Erwinia tasmaniensis",
"Methanosaeta thermophila PT",
"Thermotoga petrophila RKU-1",
"Chlamydophila pneumoniae AR39",
"Alkalilimnicola ehrlichei MLHE-1",
"Roseobacter denitrificans OCh 114",
"Candidatus Desulforudis audaxviator MP104C",
"Gluconobacter oxydans 621H",
"Campylobacter curvus 525.92",
"Clostridium beijerinckii NCIMB 8052",
"Clostridium novyi NT",
"Lactobacillus brevis ATCC 367",
"Streptococcus sanguinis SK36",
"Frankia alni ACN14a",
"Clostridium thermocellum ATCC 27405",
"Desulfotomaculum reducens MI-1",
"Opitutus terrae PB90-1",
"Methanoculleus marisnigri JR1",
"Flavobacterium psychrophilum JIP02/86",
"Brucella ovis ATCC 25840",
"Pyrococcus furiosus DSM 3638",
"Candidatus Ruthia magnifica str. Cm (Calyptogena magnifica)",
"Caulobacter crescentus CB15",
"Idiomarina loihiensis L2TR",
"Propionibacterium acnes KPA171202",
"Polaromonas naphthalenivorans CJ2",
"Psychromonas ingrahamii 37",
"Bradyrhizobium japonicum USDA 110",
"Methanothermobacter thermautotrophicus str. Delta H",
"Salinispora tropica CNB-440",
"Acinetobacter baumannii ATCC 17978",
"Ralstonia solanacearum GMI1000",
"Acholeplasma laidlawii PG-8A",
"Alcanivorax borkumensis SK2",
"Methanosphaera stadtmanae DSM 3091",
"Clostridium tetani E88",
"Bordetella petrii DSM 12804",
"Thermoplasma acidophilum DSM 1728",
"Wolbachia endosymbiont of Drosophila melanogaster",
"Chromohalobacter salexigens DSM 3043",
"Actinobacillus succinogenes 130Z",
"Neisseria gonorrhoeae FA 1090",
"Mesoplasma florum L1",
"Chlamydia trachomatis D/UW-3/CX",
"Mycoplasma pulmonis UAB CTIP",
"Bordetella avium 197N",
"Dinoroseobacter shibae DFL 12",
"Shewanella sediminis HAW-EB3",
"Leptospira biflexa serovar Patoc strain 'Patoc 1 (Paris)'",
"Leptothrix cholodnii SP-6",
"Ochrobactrum anthropi ATCC 49188",
"Pelodictyon luteolum DSM 273",
"Chlamydophila caviae GPIC",
"Burkholderia vietnamiensis G4",
"Corynebacterium diphtheriae NCTC 13129",
"Desulfotalea psychrophila LSv54",
"Nitrosopumilus maritimus SCM1",
"Rhodococcus sp. RHA1",
"Klebsiella pneumoniae subsp. pneumoniae MGH 78578",
"Jannaschia sp. CCS1",
"Arcobacter butzleri RM4018",
"Arthrobacter sp. FB24",
"Exiguobacterium sibiricum 255-15",
"Dichelobacter nodosus VCS1703A",
"Deinococcus radiodurans R1",
"Mycobacterium bovis BCG str. Pasteur 1173P2",
"Acinetobacter sp. ADP1",
"Haloarcula marismortui ATCC 43049",
"Rickettsia canadensis str. McKiel",
"Methylobacterium extorquens PA1",
"Dehalococcoides sp. CBDB1",
"Borrelia burgdorferi B31",
"Bacillus anthracis str. 'Ames Ancestor'",
"Campylobacter jejuni RM1221",
"Archaeoglobus fulgidus DSM 4304",
"Enterobacter sp. 638",
"Streptococcus gordonii str. Challis substr. CH1",
"Anaplasma phagocytophilum HZ",
"Methylococcus capsulatus str. Bath",
"Hyperthermus butylicus DSM 5456",
"Desulfovibrio desulfuricans subsp. desulfuricans str. G20",
"Thermoplasma volcanium GSS1",
"Candidatus Korarchaeum cryptofilum OPF8",
"Bacillus thuringiensis str. Al Hakam",
"Vibrio vulnificus YJ016",
"Verminephrobacter eiseniae EF01-2",
"Shewanella halifaxensis HAW-EB4",
"Streptococcus mutans UA159",
"Desulfovibrio vulgaris subsp. vulgaris str. Hildenborough",
"Methanospirillum hungatei JF-1",
"Orientia tsutsugamushi str. Boryong",
"Bacteroides fragilis YCH46",
"Pyrobaculum islandicum DSM 4184",
"Psychrobacter cryohalolentis K5",
"Methylobacterium sp. 4-46",
"Serratia proteamaculans 568",
"Saccharopolyspora erythraea NRRL 2338",
"Streptococcus agalactiae NEM316",
"Renibacterium salmoninarum ATCC 33209",
"Zymomonas mobilis subsp. mobilis ZM4",
"Magnetospirillum magneticum AMB-1",
"Rickettsia massiliae MTU5",
"Rhizobium leguminosarum bv. viciae 3841",
"Yersinia pseudotuberculosis IP 31758",
"Listeria innocua Clip11262",
"Pelobacter carbinolicus DSM 2380",
"Clostridium acetobutylicum ATCC 824",
"Methanosarcina barkeri str. Fusaro",
"Candidatus Carsonella ruddii PV",
"Bacillus halodurans C-125",
"Streptomyces coelicolor A3(2)",
"Xylella fastidiosa 9a5c",
"Citrobacter koseri ATCC BAA-895",
"Staphylococcus saprophyticus subsp. saprophyticus ATCC 15305",
"Finegoldia magna ATCC 29328",
"Rhodopseudomonas palustris BisA53",
"Sphingomonas wittichii RW1",
"Bartonella quintana str. Toulouse",
"Caldivirga maquilingensis IC-167",
"Nitrosospira multiformis ATCC 25196",
"Sulfolobus solfataricus P2",
"Rhizobium etli CFN 42",
"Mycobacterium ulcerans Agy99",
"Polynucleobacter sp. QLW-P1DMWA-1",
"Haemophilus ducreyi 35000HP",
"Nitrosomonas eutropha C91",
"Vibrio cholerae O395",
"Thiobacillus denitrificans ATCC 25259",
"Pyrobaculum aerophilum str. IM2",
"Rhodospirillum rubrum ATCC 11170",
"Sulfurihydrogenibium sp. YO3AOP1",
"Shigella boydii CDC 3083-94",
"Rhodobacter sphaeroides 2.4.1",
"Photorhabdus luminescens subsp. laumondii TTO1",
"Mycoplasma genitalium G37",
"Clostridium phytofermentans ISDg",
"Candidatus Vesicomyosocius okutanii HA",
"Nitrosococcus oceani ATCC 19707",
"Sinorhizobium medicae WSM419",
"Brucella suis 1330",
"Herminiimonas arsenicoxydans",
"Leptospira borgpetersenii serovar Hardjo-bovis L550",
"Anaplasma marginale str. St. Maries",
"Methanococcus maripaludis C5",
"Brucella melitensis 16M",
"Shewanella putrefaciens CN-32",
"Caldicellulosiruptor saccharolyticus DSM 8903",
"Thermofilum pendens Hrk 5",
"Mycobacterium smegmatis str. MC2 155",
"Salmonella typhimurium LT2",
"Bacillus licheniformis ATCC 14580",
"Clostridium perfringens ATCC 13124",
"Janthinobacterium sp. Marseille",
"Bordetella parapertussis 12822",
"Petrotoga mobilis SJ95",
"Lactobacillus casei ATCC 334",
"Alkaliphilus oremlandii OhILAs",
"Bacillus amyloliquefaciens FZB42",
"Sinorhizobium meliloti 1021",
"Salinibacter ruber DSM 13855",
"Lactobacillus gasseri ATCC 33323",
"Rubrobacter xylanophilus DSM 9941",
"Burkholderia multivorans ATCC 17616",
"Colwellia psychrerythraea 34H",
"Xanthomonas oryzae pv. oryzae MAFF 311018",
"Kineococcus radiotolerans SRS30216",
"Onion yellows phytoplasma OY-M",
"Yersinia enterocolitica subsp. enterocolitica 8081",
"Pasteurella multocida subsp. multocida str. Pm70",
"Chromobacterium violaceum ATCC 12472",
"Mesorhizobium loti MAFF303099",
"Burkholderia xenovorans LB400",
"Acidobacteria bacterium Ellin345",
"Francisella philomiragia subsp. philomiragia ATCC 25017",
"Treponema pallidum subsp. pallidum str. Nichols",
"Clostridium botulinum A3 str. Loch Maree",
"Pseudomonas syringae pv. tomato str. DC3000",
"Pyrococcus horikoshii OT3",
"Helicobacter hepaticus ATCC 51449",
"Thermoproteus neutrophilus V24Sta",
"Thiomicrospira crunogena XCL-2",
"Helicobacter pylori 26695",
"Tropheryma whipplei str. Twist",
"Candidatus Blochmannia pennsylvanicus str. BPEN",
"Shewanella sp. ANA-3",
"Xanthobacter autotrophicus Py2",
"Burkholderia phymatum STM815",
"Rickettsia rickettsii str. 'Sheila Smith'",
"Mycoplasma gallisepticum R",
"Bifidobacterium longum DJO10A",
"Cytophaga hutchinsonii ATCC 33406",
"Erythrobacter litoralis HTCC2594",
"Mycoplasma penetrans HF-2",
"Carboxydothermus hydrogenoformans Z-2901",
"Synechococcus elongatus PCC 7942",
"Methanosarcina mazei Go1",
"Thermotoga sp. RQ2",
"Vibrio fischeri ES114",
"Methylobacterium radiotolerans JCM 2831",
"Rhodoferax ferrireducens T118",
"Vibrio parahaemolyticus RIMD 2210633",
"Chlamydia muridarum Nigg",
"Shigella flexneri 2a str. 301",
"Novosphingobium aromaticivorans DSM 12444",
"Nocardia farcinica IFM 10152",
"Gloeobacter violaceus PCC 7421",
"Streptomyces avermitilis MA-4680",
"Azorhizobium caulinodans ORS 571",
"Parabacteroides distasonis ATCC 8503",
"Staphylothermus marinus F1",
"Sphingopyxis alaskensis RB2256",
"Methanocaldococcus jannaschii DSM 2661",
"Bordetella pertussis Tohama I",
"Streptococcus pyogenes MGAS10750",
"Corynebacterium jeikeium K411",
"Psychrobacter sp. PRwf-1",
"Lactobacillus helveticus DPC 4571",
"Legionella pneumophila str. Paris",
"Candidatus Protochlamydia amoebophila UWE25",
"Methylobacillus flagellatus KT",
"Methanococcoides burtonii DSM 6242",
"Symbiobacterium thermophilum IAM 14863",
"Treponema denticola ATCC 35405",
"Campylobacter hominis ATCC BAA-381",
"Mycoplasma mycoides subsp. mycoides SC str. PG1",
"Wolinella succinogenes DSM 1740",
"Rickettsia felis URRWXCal2",
"Moorella thermoacetica ATCC 39073",
"Bartonella bacilliformis KC583"
]

for s in species_list:
	get_genome(s)

